src/exporter/DaemonMetricCollector.cc

```diff
@@ -102,62 +102,114 @@ std::string quote(std::string value) { return "\"" + value + "\""; }
+                        << daemon_name << dendl;
+                continue;
+              }
+              labels.insert(extra_labels.begin(), extra_labels.end());
+
+              // For now this is only required for rgw multi-site metrics
+              auto multisite_labels_and_name = add_fixed_name_metrics(counter_name);
+              if (!multisite_labels_and_name.first.empty()) {
+                labels.insert(multisite_labels_and_name.first.begin(),
+                              multisite_labels_and_name.first.end());
+                counter_name = multisite_labels_and_name.second;
+              }
+              auto perf_values = counters_values.at(counter_name_init);
+              dump_asok_metric(counter_group, perf_values, counter_name, labels);
+            } catch (const std::exception &e) {
+              dout(1) << "Exception in counter processing for " << daemon_name << ": " << e.what() << dendl;
+              continue;
+            }
+          }
+        } catch (const std::exception &e) {
+          dout(1) << "Exception in schema/dump iteration for " << daemon_name << ": " << e.what() << dendl;
           continue;
         }
-        labels.insert(extra_labels.begin(), extra_labels.end());
-
-        // For now this is only required for rgw multi-site metrics
-        auto multisite_labels_and_name = add_fixed_name_metrics(counter_name);
-        if (!multisite_labels_and_name.first.empty()) {
-          labels.insert(multisite_labels_and_name.first.begin(),
-                        multisite_labels_and_name.first.end());
-          counter_name = multisite_labels_and_name.second;
-        }
-        auto perf_values = counters_values.at(counter_name_init);
-        dump_asok_metric(counter_group, perf_values, counter_name, labels);
       }
     }
+  } catch (const std::exception &e) {
+    dout(1) << "Exception in parse_asok_metrics for " << daemon_name << ": " << e.what() << dendl;
+    return;
   }
 }
 
+/*
+perf_values can be either a int/double or a json_object. Since
+   json_value is a wrapper of both we use that class.
+ */
+void DaemonMetricCollector::dump_asok_metric(json_object perf_info,
+                                             json_value perf_values,
+                                             std::string name,
+                                             labels_t labels) {
+  try {
+    if (!perf_info.if_contains("type") ||
+        !perf_info.if_contains("metric_type") ||
+        !perf_info.if_contains("description")) {
+      dout(1) << "Missing required key in perf_info for metric: " << name << dendl;
+      return;
+    }
+    int64_t type = perf_info["type"].as_int64();
+    std::string metric_type =
+        boost_string_to_std(perf_info["metric_type"].as_string());
+    std::string description =
+        boost_string_to_std(perf_info["description"].as_string());
+
+    if (type & PERFCOUNTER_LONGRUNAVG) {
+      int64_t count = perf_values.as_object()["avgcount"].as_int64();
```

Comment:
I will check explicitly for the existence of the "avgcount" and "sum" attributes before doing nothing with them .

Something like:

```
const auto& metric = perf_values.as_object();

if (metric.contains("avgcount")) {
    int64_t count = metric.at("avgcount").as_int64();
    add_metric(builder, count, name + "_count", description + " Count", "counter", labels);
} else {
   dout(1) << "avgcount attribute not available for " << name << "metric" << denl 
}
```

Reply:
Done